### PR TITLE
Replace internally developed CSR GEMM with a call to MKL

### PR DESCRIPTION
**Pull request by Vika-F**
- internally developed GEMM was removed
- call to MKL's sparse GEMM was added into clusters assignment step in sparse K-means (see `assign_clusters` kernel)
- incorrect use of communicator was removed
- Sparse method was aligned with dense method. Now both methods give the same results on the same input data passed in sparse and dense layout respectively

### CODE REVIEW

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_csr_impl.hpp [hunk 1]
```diff
@@ -49,91 +52,83 @@ sycl::event compute_data_squares(sycl::queue& q,
...
-                                                Float(0),
-                                                sycl::ext::oneapi::plus<Float>());
-                    if (local_id == 0) {
-                        res_ptr[row_idx * b_row_count + col_idx] =
-                            beta * res_ptr[row_idx * b_row_count + col_idx] + alpha * scalar_mul;
-                    }
-                }
-            }
+        cgh.parallel_for(range, [=](auto item) {
+            auto i = item[0];
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_csr_impl.hpp:91]: const std::int64_t i = item[0], j = item[1]; is probably better

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/infer_kernel_dpc.cpp [hunk 2]
```diff
@@ -154,23 +158,21 @@ struct infer_kernel_gpu<Float, method::lloyd_csr, task::clustering> {
...
                                             distances,
                                             arr_responses,
                                             arr_closest_distances,
                                             { data_squares_event, centroid_squares_event });
+
         auto objective_function =
             calc_objective_function(queue, arr_closest_distances, { assign_event });
-        {
-            // Reduce objective function value over all ranks
-            comm.allreduce(objective_function).wait();
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/kmeans/backend/gpu/infer_kernel_dpc.cpp:172]: whats the reason of removing this comm operation?

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_csr_impl.hpp [hunk 3]
```diff
@@ -142,77 +137,47 @@ sycl::event assign_clusters(sycl::queue& q,
...
-                                            sycl::ext::oneapi::minimum<std::int32_t>());
-                if (local_id == 0) {
-                    responses_ptr[row_idx] = closest_id;
-                    closest_dists_ptr[row_idx] = closest;
-                }
-            }
-        });
-    });
-    return event;
+    // Workaround. Sparce gemm cannot accept transposed dense inputs for now.
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_csr_impl.hpp:141]: Please mark it as TODO for for more easy tracking

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_csr_dpc.cpp [hunk 4]
```diff
@@ -206,11 +223,6 @@ struct train_kernel_gpu<Float, method::lloyd_csr, task::clustering> {
             calc_objective_function(queue,
                                     arr_closest_distances,
                                     { last_event, centroid_squares_event, assign_event });
-        {
-            // Reduce objective function value over all ranks
-            auto obj_func_reduce_event = comm.allreduce(objective_function);
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_csr_dpc.cpp:211]: I didnt find any replacement for comm usage, can you please clarify why it works in another way now

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_csr_impl.hpp [hunk 5]
```diff
@@ -49,91 +52,83 @@ sycl::event compute_data_squares(sycl::queue& q,
...
-                                                acc,
-                                                Float(0),
-                                                sycl::ext::oneapi::plus<Float>());
-                    if (local_id == 0) {
-                        res_ptr[row_idx * b_row_count + col_idx] =
-                            beta * res_ptr[row_idx * b_row_count + col_idx] + alpha * scalar_mul;
-                    }
-                }
-            }
+        cgh.parallel_for(range, [=](auto item) {
```
[the remark this answers is not in the document]
Comment (Alexandr-Solovev) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_csr_impl.hpp:90]: and also makes sense to init range here

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_csr_impl.hpp [hunk 6]
```diff
@@ -256,157 +224,108 @@ sycl::event update_centroids(sycl::queue& q,
...
+    // Compute sum of data points belonging to each centroid
     auto centroids_sum_event = q.submit([&](sycl::handler& cgh) {
         cgh.depends_on(clean_event);
-        local_accessor_rw_t<Float> local_centroid(column_count, cgh);
-        cgh.parallel_for(range, [=](auto it) {
-            const auto cluster_id = it.get_global_id(0);
-            const auto row_shift = it.get_global_id(1);
-            const auto local_id = static_cast<std::int64_t>(it.get_local_id(2));
-            if (counts_ptr[cluster_id] == 0) {
+        cgh.depends_on(deps);
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_csr_impl.hpp:250]: you can unite it with the previous depends_on

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/empty_cluster_handling.hpp [hunk 7]
```diff
@@ -17,17 +17,37 @@
 #pragma once
 
 #include "oneapi/dal/backend/communicator.hpp"
+#include "oneapi/dal/backend/interop/common_dpc.hpp"
```
[the remark this answers is not in the document]
Comment (Vika-F) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/empty_cluster_handling.hpp]: Modified the code to make this not necessary.

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/empty_cluster_handling_dpc.cpp [hunk 8]
```diff
@@ -97,7 +98,6 @@ static auto fill_empty_cluster_indices(sycl::queue& queue,
         if (host_counters_ptr[i] > 0) {
             continue;
         }
-
```
[the remark this answers is not in the document]
Comment (Vika-F) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/empty_cluster_handling_dpc.cpp]: Reverted.

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/empty_cluster_handling_dpc.cpp [hunk 9]
```diff
@@ -148,6 +148,49 @@ static auto copy_candidates_from_data(sycl::queue& queue,
...
+    Float* centroids_ptr = centroids.get_mutable_data();
+    const std::int32_t* candidate_indices_ptr = candidates.get_indices().get_data();
+    const std::int32_t* empty_cluster_indices_ptr =
+        candidates.get_empty_cluster_indices().get_data();
+
+    auto event = queue.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(deps);
+
+        const auto range = sycl::range(candidate_count);
+        cgh.parallel_for(range, [=](auto id) {
```
[the remark this answers is not in the document]
Comment (Vika-F) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/empty_cluster_handling_dpc.cpp]: Done.

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/empty_cluster_handling_dpc.cpp [hunk 10]
```diff
@@ -148,6 +148,49 @@ static auto copy_candidates_from_data(sycl::queue& queue,
...
+            const std::int64_t dst_i = empty_cluster_indices_ptr[id];
+            const std::int64_t src_i = candidate_indices_ptr[id];
+            const std::int64_t begin_idx = row_offsets_ptr[src_i];
+            const std::int64_t end_idx = row_offsets_ptr[src_i + 1];
+            for (auto j = 0; j < column_count; ++j) {
```
[the remark this answers is not in the document]
Comment (Vika-F) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/empty_cluster_handling_dpc.cpp]: Yes, changed to `std::int64_t`

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_csr_impl.hpp (hunk 1 above)
Comment (Vika-F) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_csr_impl.hpp]: Done.

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_csr_impl.hpp (hunk 3 above)
Comment (Vika-F) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_csr_impl.hpp:140]: Distributed implementation for CSR method is incomplete and is not functional. Removing for now.

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/infer_kernel_dpc.cpp (hunk 2 above)
Comment (Vika-F) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/infer_kernel_dpc.cpp:172]: Distributed implementation for CSR method is incomplete and is not functional. Removing for now.

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_csr_impl.hpp (hunk 6 above)
Comment (Vika-F) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_csr_impl.hpp]: Done.

**Review by Alexandr-Solovev — APPROVED** [review]

**Comment by Vika-F** [comment]
Re-run pre-commit: https://intel-ci.intel.com/ef3aa431-06ee-f183-af2d-a4bf010d0e2e

**Comment by Vika-F** [comment]
The work was moved to #2959

**Pull request closed without merging**